libs/optimizer/src/lib/optimizer.ts

```diff
@@ -0,0 +1,348 @@
+  const result: number[] = [];
+  pushInResult(getEntryPointsStartIndex(solution, originalTrack, solverTrack));
+  pushInResult(getClosingPointsInIndex(solution, originalTrack, solverTrack));
+  solution.scoreInfo?.tp?.map((turnPoint) => turnPoint.r)
+    .forEach(index => pushInResult(getPointIndex(index, originalTrack, solverTrack)));
+  pushInResult(getClosingPointsOutIndex(solution, originalTrack, solverTrack));
+  pushInResult(getEntryPointsFinishIndex(solution, originalTrack, solverTrack));
+  return result;
+
+  function pushInResult(index: number) {
+    if (index >= 0) {
+      result.push(index);
+    }
+  }
+}
+
+function getEntryPointsStartIndex(solution: Solution, originalTrack: ScoringTrack, solverTrack: ScoringTrack): number {
+  console.debug("getEntryPointsStartIndex", solution.scoreInfo?.ep?.start.r);
+  return getPointIndex(solution.scoreInfo?.ep?.start.r, originalTrack, solverTrack);
+}
+
+function getClosingPointsInIndex(solution: Solution, originalTrack: ScoringTrack, solverTrack: ScoringTrack): number {
+  console.debug("getClosingPointsInIndex", solution.scoreInfo?.cp?.in.r);
+  return getPointIndex(solution.scoreInfo?.cp?.in.r, originalTrack, solverTrack);
+}
+
+function getClosingPointsOutIndex(solution: Solution, originalTrack: ScoringTrack, solverTrack: ScoringTrack): number {
+  console.debug("getClosingPointsOutIndex", solution.scoreInfo?.cp?.out.r);
+  return getPointIndex(solution.scoreInfo?.cp?.out.r, originalTrack, solverTrack);
+}
+
+function getEntryPointsFinishIndex(solution: Solution, originalTrack: ScoringTrack, solverTrack: ScoringTrack): number {
+  console.debug("getEntryPointsFinishIndex", solution.scoreInfo?.ep?.finish.r);
+  return getPointIndex(solution.scoreInfo?.ep?.finish.r, originalTrack, solverTrack);
+}
+
+
+function getPointIndex(
+  index: number | undefined,
+  originalTrack: ScoringTrack,
+  solverTrack: ScoringTrack,
+): number {
+  if (index === undefined) {
+    return -1;
+  }
+  return solutionContainsValidIndices(originalTrack, solverTrack)
+    ? index
+    : getIndexInOriginalTrack(index, originalTrack, solverTrack);
+}
+
+function solutionContainsValidIndices(originalTrack: ScoringTrack, solverTrack: ScoringTrack): boolean {
+  return originalTrack.points.length === solverTrack.points.length;
+}
+
+
+function getIndexInOriginalTrack(
+  index: number,
+  originalTrack: ScoringTrack,
+  solverTrack: ScoringTrack,
+): number {
+  const solutionPoint = solverTrack.points[index];
+  let indexInOriginalTrack = -1;
+  let closestDistance = Number.MAX_VALUE;
+  for (let i = 0; i < originalTrack.points.length; i++) {
+    let point = originalTrack.points[i];
+    const distance = getDistance(point, solutionPoint);
+    if (distance < closestDistance) {
+      closestDistance = distance;
+      indexInOriginalTrack = i;
+    }
+  }
+  return indexInOriginalTrack;
+}
+
+// another implementation of deep copy to avoid dependency from @flyxc/common
```

Comment:
Good idea, you can drop this line so that there is no dependency in the comments

Reply:
Ok. Changed